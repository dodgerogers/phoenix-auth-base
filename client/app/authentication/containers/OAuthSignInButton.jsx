import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { oAuthSignIn } from 'redux-oauth';


const OAuthSignInButton = (props) => {
  const handleClick = () => {
    const { provider } = props;
    props.oAuthSignIn({ provider });
  };

  return (
    <Button
      fluid={props.fluid}
      color={props.color}
      loading={props.loading}
      disabled={props.disabled}
      onClick={handleClick}
    >
      {props.children}
    </Button>
  );
}

OAuthSignInButton.propTypes = {
  oAuthSignIn: PropTypes.func.isRequired,
  provider: PropTypes.string.isRequired,
  children: PropTypes.node,
  dispatch: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  fluid: PropTypes.bool,
  color: PropTypes.string,
};

OAuthSignInButton.defaultProps = {
  children: <span>OAuth Sign In</span>,
};

function mapStateToProps({ auth }, ownProps) {
  return {
    disabled: auth.getIn(['user', 'isSignedIn']),
    loading: auth.getIn(['oAuthSignIn', ownProps.provider, 'loading']),
  }
}

export { OAuthSignInButton as PureComponent };
export default connect(mapStateToProps, { oAuthSignIn })(OAuthSignInButton);
