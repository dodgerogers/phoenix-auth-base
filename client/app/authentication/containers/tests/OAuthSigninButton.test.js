import React from 'react';
import { mount, shallow } from 'enzyme';
import { fromJS } from 'immutable';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Button } from 'semantic-ui-react';
import OAuthSignInButton, { PureComponent } from '../OAuthSignInButton';

const mockStore = configureStore([thunk]);

describe('OmniautSiginButton', () => {
  const provider = "facebook";
  const fullProps = {
    oAuthSignIn: jest.fn(),
    provider
  };
  const initialState = (isSignedIn = false, oAuthSignIn = false) => ({
    auth: fromJS({
      user: {
        isSignedIn: isSignedIn,
      },
      oAuthSignIn: {
        facebook: {
          loading: oAuthSignIn,
        }
      }
    }),
  });

  const wrapper = (props, store) => mount(
    <Provider store={store}>
      <OAuthSignInButton provider={provider}>
        Sign in
      </OAuthSignInButton>
    </Provider>
  );

  const matchesSnapshotBasedOnProps = (initialState) => {
    const store = mockStore(initialState);
    const component = wrapper(fullProps, store);

    expect(toJson(component)).toMatchSnapshot();
  }

  it('matches snapshot when is neither disabled nor loading', () => {
    matchesSnapshotBasedOnProps(initialState());
  });

  it('matches snapshot when is disabled', () => {
    matchesSnapshotBasedOnProps(initialState(true, false));
  });

  it('matches snapshot when is loading', () => {
    matchesSnapshotBasedOnProps(initialState(false, true));
  });

  it('onClick', () => {
    const store = mockStore(initialState());
    const component = shallow(<PureComponent {...fullProps} />);

    component.find(Button).simulate('click');

    expect(fullProps.oAuthSignIn).toHaveBeenCalledWith({ provider });
  });
});
