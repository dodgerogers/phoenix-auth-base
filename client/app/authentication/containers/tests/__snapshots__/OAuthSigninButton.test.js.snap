// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OmniautSiginButton matches snapshot when is disabled 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(OAuthSignInButton)
    provider="facebook"
  >
    <OAuthSignInButton
      disabled={true}
      loading={false}
      oAuthSignIn={[Function]}
      provider="facebook"
    >
      <Button
        as="button"
        disabled={true}
        loading={false}
        onClick={[Function]}
      >
        <button
          className="ui disabled button"
          disabled={true}
          onClick={[Function]}
          tabIndex={-1}
        >
          Sign in
        </button>
      </Button>
    </OAuthSignInButton>
  </Connect(OAuthSignInButton)>
</Provider>
`;

exports[`OmniautSiginButton matches snapshot when is loading 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(OAuthSignInButton)
    provider="facebook"
  >
    <OAuthSignInButton
      disabled={false}
      loading={true}
      oAuthSignIn={[Function]}
      provider="facebook"
    >
      <Button
        as="button"
        disabled={false}
        loading={true}
        onClick={[Function]}
      >
        <button
          className="ui loading button"
          onClick={[Function]}
        >
          Sign in
        </button>
      </Button>
    </OAuthSignInButton>
  </Connect(OAuthSignInButton)>
</Provider>
`;

exports[`OmniautSiginButton matches snapshot when is neither disabled nor loading 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(OAuthSignInButton)
    provider="facebook"
  >
    <OAuthSignInButton
      disabled={false}
      loading={false}
      oAuthSignIn={[Function]}
      provider="facebook"
    >
      <Button
        as="button"
        disabled={false}
        loading={false}
        onClick={[Function]}
      >
        <button
          className="ui button"
          onClick={[Function]}
        >
          Sign in
        </button>
      </Button>
    </OAuthSignInButton>
  </Connect(OAuthSignInButton)>
</Provider>
`;
